apollo-router/src/plugins/cache/entity.rs

```diff
@@ -1507,37 +1520,136 @@ fn get_entity_keys_from_supergraph_schema(
+        if field.selection_set.is_empty() {
+            let value = representation.remove(field.name.as_str())?;
+            // `value` must be a scalar.
+            if matches!(value, Value::Object(_)) {
+                return None;
+            }
+            // Move the scalar field to the `result`.
+            result.insert(ByteString::from(field.name.as_str()), value);
+            continue;
+        } else {
+            let value = representation.get_mut(field.name.as_str())?;
+            // Update the sub-selection set.
+            let Value::Object(sub_value) = value else {
+                return None;
+            };
+            let removed = take_selection_set(sub_value, &field.selection_set)?;
+            result.insert(
+                ByteString::from(field.name.as_str()),
+                Value::Object(removed),
+            );
+        }
+    }
+    Some(result)
+}
+
+// The inverse of `take_selection_set`.
+fn merge_representation(
+    dest: &mut serde_json_bytes::Map<ByteString, Value>,
+    source: serde_json_bytes::Map<ByteString, Value>,
+) {
+    source.into_iter().for_each(|(key, src_value)| {
+        // Note: field sets can't have aliases.
+        let Some(dest_value) = dest.get_mut(&key) else {
+            dest.insert(key, src_value);
+            return;
+        };
+
+        // Overlapping fields must be objects.
+        if let (Value::Object(dest_sub_value), Value::Object(src_sub_value)) =
+            (dest_value, src_value)
+        {
+            // Merge sub-values
+            merge_representation(dest_sub_value, src_sub_value);
+        }
+    });
+}
 
-    Ok(entity_keys)
+// Order-insensitive structural hash of the representation value
+pub(crate) fn hash_representation(
+    representation: &serde_json_bytes::Map<ByteString, Value>,
+) -> String {
+    let mut digest = Sha256::new();
+    fn hash(state: &mut Sha256, fields: &serde_json_bytes::Map<ByteString, Value>) {
+        fields
+            .iter()
+            .sorted_by(|a, b| a.0.cmp(b.0))
+            .for_each(|(k, v)| {
+                state.update(serde_json::to_string(k).unwrap().as_bytes());
+                state.update(":".as_bytes());
+                match v {
+                    serde_json_bytes::Value::Object(obj) => {
+                        state.update("{".as_bytes());
+                        hash(state, obj);
+                        state.update("}".as_bytes());
+                    }
+                    _ => state.update(serde_json::to_string(v).unwrap().as_bytes()),
+                }
+            });
+    }
+    hash(&mut digest, representation);
+    hex::encode(digest.finalize().as_slice())
 }
 
 // Only hash the list of entity keys
 pub(crate) fn hash_entity_key(
-    entity_keys: &IndexMap<ByteString, serde_json_bytes::Value>,
+    entity_keys: &serde_json_bytes::Map<ByteString, serde_json_bytes::Value>,
 ) -> String {
+    #[cfg(test)]
+    dbg!(&entity_keys);
```

Comment:
Maybe, remove this?

Reply:
Actually I would like to keep it, it's pretty easy when running the test to debug if anything goes wrong

Reply:
I think this should not stay as-is because it’d be a hindrance for debugging a test if something _else_ (unrelated to this function) goes wrong. If you want to keep some logging in `dev` it should be opt-in somehow. Maybe with `tracing::trace!`?

Reply:
Replaced with `tracing::trace!`